File: C++/ps_baekjoon/2109/main.cpp
#include <bits/stdc++.h>
using namespace std;

struct lecture
{
    int day, pay;
    lecture() {}
    bool operator<(const lecture& obj) const { return this->pay < obj.pay; }
};

int main()
{
    cin.tie(0)->sync_with_stdio(0);
    int n, max_day = 0, res = 0;
    cin >> n;
    vector<lecture> lectures(n);
    priority_queue<lecture> pq;
    for (int i = 0; i < n; ++i)
    {
        cin >> lectures[i].pay >> lectures[i].day;
        max_day = max(max_day, lectures[i].day);
    }
    sort(lectures.begin(), lectures.end(), [](lecture a, lecture b) { return a.day < b.day; });
    for (int today = max_day; today > 0; --today)
    {
        while (!lectures.empty() && lectures.back().day == today)
        {
            pq.push(lectures.back());
            lectures.pop_back();
        }
        if (!pq.empty())
        {
            res += pq.top().pay;
            pq.pop();
        }
    }
    cout << res;
    return 0;
}

File: Python/ps_baekjoon/2109/main.py
import sys
from queue import PriorityQueue

readline, write = (sys.stdin.readline, sys.stdout.write)
pq = PriorityQueue()
res = 0


class lecture:
    __slots__ = ["pay", "day"]

    def __init__(self, pay, day):
        self.pay = pay
        self.day = day

    def __lt__(self, obj):
        return self.pay > obj.pay


n = int(readline())
lectures = [lecture(*[int(i) for i in readline().split()][:2]) for _ in range(n)]
lectures.sort(key=lambda x: x.day)
max_day = lectures[-1].day if lectures else 0
for today in range(max_day, 0, -1):
    while lectures and lectures[-1].day == today:
        pq.put(lectures.pop())
    if not pq.empty():
        res += pq.get().pay
write(f"{res}")
